### PR TITLE
test(RHINENG-25382): unit test for useBulkSelect

### PR DESCRIPTION
**Pull request by computercamplove**
## Jira
https://redhat.atlassian.net/browse/RHINENG-25382

## What
unit test for `useBulkSelect`

## Why
require to release SystemsView to prod-stable

## How
all critical paths remain fully tested:
  - state calculations (empty, partial, full selection)
  - all `onBulkSelect` behaviors across different sources and values
  - edge cases and real-world scenarios
  - callback memoization

## Summary by Sourcery

Add comprehensive unit test coverage for the useBulkSelect hook in SystemsView, exercising selection state, bulk selection behaviors, and edge cases.

Tests:
- Add extensive useBulkSelect hook tests covering empty, partial, full, and cross-page selection states.
- Verify onBulkSelect handling for all BulkSelectValue options, sources, and unknown inputs.
- Test edge and real-world scenarios including empty data, total mismatches, mixed-page selections, and rapid successive calls.

### CODE REVIEW

File: src/components/SystemsView/hooks/useBulkSelect.test.ts [hunk 1]
```diff
@@ -0,0 +1,404 @@
...
+};
+
+const createTestItems = (count: number): TestItem[] =>
+  Array.from({ length: count }, (_, i) => ({
+    id: `item-${i}`,
+    name: `Item ${i}`,
+  }));
+
+describe('useBulkSelect', () => {
+  describe('fundamental behaviors', () => {
```
Comment (raswonders) [inline src/components/SystemsView/hooks/useBulkSelect.test.ts:36]: we are checking isPageSelected and isPartiallySelected state combinations, but we  never check isFullySelected in here.

File: src/components/SystemsView/hooks/useBulkSelect.test.ts [hunk 2]
```diff
@@ -0,0 +1,404 @@
...
+          useBulkSelect({ selection, rows, total: 5 }),
+        );
+
+        await result.current.onBulkSelect(BulkSelectValue.page, 'checkbox');
+
+        expect(selection.onSelect).toHaveBeenCalledWith(true, rows);
+        expect(selection.setSelected).not.toHaveBeenCalled();
+      });
+
+      it('does nothing when source is not dropdown or checkbox', async () => {
```
Comment (raswonders) [inline src/components/SystemsView/hooks/useBulkSelect.test.ts]: What guarantees this test gives us? To me it doesn't seem very helpful.

File: src/components/SystemsView/hooks/useBulkSelect.test.ts [hunk 3]
```diff
@@ -0,0 +1,404 @@
...
+          BulkSelectValue.page,
+          'unknown' as BulkSelectSource,
+        );
+
+        expect(selection.onSelect).not.toHaveBeenCalled();
+        expect(selection.setSelected).not.toHaveBeenCalled();
+      });
+    });
+
+    describe('other/unknown values', () => {
```
Comment (raswonders) [inline src/components/SystemsView/hooks/useBulkSelect.test.ts]: We don't need this as TS makes sure call sites only ever use BulkSelectValue 

Image

File: src/components/SystemsView/hooks/useBulkSelect.test.ts [hunk 4]
```diff
@@ -0,0 +1,404 @@
...
+        await result.current.onBulkSelect('unknown' as BulkSelectValue);
+
+        expect(selection.onSelect).not.toHaveBeenCalled();
+        expect(selection.setSelected).not.toHaveBeenCalled();
+      });
+    });
+  });
+
+  describe('edge cases', () => {
+    it('handles empty rows array', () => {
```
Comment (raswonders) [inline src/components/SystemsView/hooks/useBulkSelect.test.ts]: this comments could be more specific. We could say how it handles the situation not just that it handles it. 

For instance it could say something: "selects no rows when selection is empty"

File: src/components/SystemsView/hooks/useBulkSelect.test.ts [hunk 5]
```diff
@@ -0,0 +1,404 @@
...
+      const { result } = renderHook(() =>
+        useBulkSelect({ selection, rows, total: 0 }),
+      );
+
+      expect(result.current.selectedCount).toBe(0);
+      expect(result.current.isPageSelected).toBe(false);
+      expect(result.current.isPartiallySelected).toBe(false);
+    });
+
+    it('handles total being 0', () => {
```
Comment (raswonders) [inline src/components/SystemsView/hooks/useBulkSelect.test.ts]: this is more of a test of empty selection than the total === 0. Empty selection is also tried in test before.

File: src/components/SystemsView/hooks/useBulkSelect.test.ts [hunk 6]
```diff
@@ -0,0 +1,404 @@
...
+
+      const { result } = renderHook(() =>
+        useBulkSelect({ selection, rows, total: 0 }),
+      );
+
+      expect(result.current.selectedCount).toBe(0);
+      expect(result.current.isPartiallySelected).toBe(false);
+    });
+
+    it('handles selection count exceeding total', () => {
```
Comment (raswonders) [inline src/components/SystemsView/hooks/useBulkSelect.test.ts:198]: Isn't this just arbitrary test can this actually happen in the code?

File: src/components/SystemsView/hooks/useBulkSelect.test.ts [hunk 7]
```diff
@@ -0,0 +1,404 @@
...
+      const { result } = renderHook(() =>
+        useBulkSelect({ selection, rows, total: 20 }),
+      );
+
+      expect(result.current.selectedCount).toBe(3);
+      expect(result.current.isPageSelected).toBe(false);
+      expect(result.current.isPartiallySelected).toBe(true);
+    });
+
+    it('correctly identifies page as selected even with items from other pages selected', () => {
```
Comment (raswonders) [inline src/components/SystemsView/hooks/useBulkSelect.test.ts]: this one maybe extra too? I guess it would be enough If we only have test checking that isPageSelected is true if whole page is selected. That should be enough.

File: src/components/SystemsView/hooks/useBulkSelect.test.ts [hunk 8]
```diff
@@ -0,0 +1,404 @@
...
+
+      const { result } = renderHook(() =>
+        useBulkSelect({ selection, rows, total: 20 }),
+      );
+
+      expect(result.current.isPageSelected).toBe(true);
+      expect(result.current.selectedCount).toBe(6);
+    });
+
+    it('updates callback when dependencies change', () => {
```
Comment (raswonders) [inline src/components/SystemsView/hooks/useBulkSelect.test.ts]: Seems like extra as well ... useCallback with its dependency array protects against this, it feels like we're testing React library here.

File: src/components/SystemsView/hooks/useBulkSelect.test.ts [hunk 9]
```diff
@@ -0,0 +1,404 @@
...
+      const { result } = renderHook(() =>
+        useBulkSelect({ selection, rows, total: 50 }),
+      );
+
+      await result.current.onBulkSelect(BulkSelectValue.page, 'checkbox');
+
+      expect(selection.setSelected).toHaveBeenCalledWith([]);
+    });
+
+    it('handles navigation between pages maintaining selections', () => {
```
Comment (raswonders) [inline src/components/SystemsView/hooks/useBulkSelect.test.ts]: this hook doesn't own selection persistence in table, that makes this test quite out of place.

**Review by raswonders — COMMENTED** [review]
Thank you for tests, coverage seem good, but I'd skip some of these as they feel redundant or not providing any real value. 

Please go over the notes let me know what you think and when you're ready for another review.

File: src/components/SystemsView/hooks/useBulkSelect.test.ts (hunk 1 above)
Comment (computercamplove) [reply src/components/SystemsView/hooks/useBulkSelect.test.ts:36]: @raswonders if we test `isPartiallySelected` when it `false`, why need to test `isFullySelected`? To test if all selected in the page we test `isPageSelected`.

File: src/components/SystemsView/hooks/useBulkSelect.test.ts (hunk 3 above)
Comment (computercamplove) [reply src/components/SystemsView/hooks/useBulkSelect.test.ts]: make sense:D thank you!

File: src/components/SystemsView/hooks/useBulkSelect.test.ts (hunk 6 above)
Comment (computercamplove) [reply src/components/SystemsView/hooks/useBulkSelect.test.ts:198]: @raswonders for example filter is applied or some selected systems are removed

File: src/components/SystemsView/hooks/useBulkSelect.test.ts (hunk 9 above)
Comment (computercamplove) [reply src/components/SystemsView/hooks/useBulkSelect.test.ts]: @raswonders gonna drop it here, but we don't have tests for `SystemsView` page chnages

File: src/components/SystemsView/hooks/useBulkSelect.test.ts (hunk 1 above)
Comment (raswonders) [reply src/components/SystemsView/hooks/useBulkSelect.test.ts:36]: Yes was wrong on this one, it struck me as weird we don't assert ever against this state but looking again I realized isFullySelected is only tracked internally. So no need.

Image

File: src/components/SystemsView/hooks/useBulkSelect.test.ts (hunk 9 above)
Comment (raswonders) [reply src/components/SystemsView/hooks/useBulkSelect.test.ts]: that's fine :) we're gonna have them

File: src/components/SystemsView/hooks/useBulkSelect.test.ts (hunk 2 above)
Comment (computercamplove) [reply src/components/SystemsView/hooks/useBulkSelect.test.ts]: @raswonders you are right - it is 2 typed. for some reason when i navigated cursor first time on `BulkSelectSource` it showed me "undefined" there too -_- but when i searched for `export type BulkSelectSource ` it started to show me right type in hints 
image

File: src/components/SystemsView/hooks/useBulkSelect.test.ts (hunk 6 above)
Comment (raswonders) [reply src/components/SystemsView/hooks/useBulkSelect.test.ts:198]: ah that's true! disregard then

File: src/components/SystemsView/hooks/useBulkSelect.test.ts (hunk 2 above)
Comment (computercamplove) [reply src/components/SystemsView/hooks/useBulkSelect.test.ts]: probably some visual code issue hint issue:)

**Comment by computercamplove** [comment]
@raswonders i applied all suggestions expect 2 you agree not to do anything. Thank you for review!

File: src/components/SystemsView/hooks/useBulkSelect.test.ts (hunk 2 above)
Comment (raswonders) [reply src/components/SystemsView/hooks/useBulkSelect.test.ts]: yeah types are always our first line of defense :)

**Review by raswonders — APPROVED** [review]
looks very good now, thanks for considering the notes!
